apiserver/facades/client/application/deployrepository.go

```diff
@@ -80,3 +140,467 @@ func validateDeployFromRepositoryArgs(_ params.DeployFromRepositoryArg) []error
+	// do get "all" we should see if there is a clean way to resolve it.
+	if resolvedOrigin.Platform.Architecture == "all" {
+		cons, err := v.state.ModelConstraints()
+		if err != nil {
+			return nil, corecharm.Origin{}, errors.Trace(err)
+		}
+		resolvedOrigin.Platform.Architecture = arch.ConstraintArch(cons, nil)
+	}
+
+	var seriesFlag string
+	if requestedOrigin.Platform.OS != "" {
+		var err error
+		seriesFlag, err = coreseries.GetSeriesFromChannel(requestedOrigin.Platform.OS, requestedOrigin.Platform.Channel)
+		if err != nil {
+			return nil, corecharm.Origin{}, errors.Trace(err)
+		}
+	}
+
+	modelCfg, err := v.model.Config()
+	if err != nil {
+		return nil, corecharm.Origin{}, errors.Trace(err)
+	}
+
+	imageStream := modelCfg.ImageStream()
+
+	workloadSeries, err := coreseries.WorkloadSeries(jujuclock.WallClock.Now(), seriesFlag, imageStream)
+	if err != nil {
+		return nil, corecharm.Origin{}, errors.Trace(err)
+	}
+
+	selector := corecharm.SeriesSelector{
+		SeriesFlag:          seriesFlag,
+		SupportedSeries:     supportedSeries,
+		SupportedJujuSeries: workloadSeries,
+		Force:               force,
+		Conf:                modelCfg,
+		FromBundle:          false,
+		Logger:              deployRepoLogger,
+	}
+
+	// Get the series to use.
+	series, err := selector.CharmSeries()
+	deployRepoLogger.Tracef("Using series %q from %v to deploy %v", series, supportedSeries, curl)
+	if charm.IsUnsupportedSeriesError(err) {
+		msg := fmt.Sprintf("%v. Use --force to deploy the charm anyway.", err)
+		if usedModelDefaultBase {
+			msg += " Used the default-series."
+		}
+		return nil, corecharm.Origin{}, errors.Errorf(msg)
+	}
+
+	var base coreseries.Base
+	if series == coreseries.Kubernetes.String() {
+		base = coreseries.LegacyKubernetesBase()
+	} else {
+		base, err = coreseries.GetBaseFromSeries(series)
+		if err != nil {
+			return nil, corecharm.Origin{}, errors.Trace(err)
+		}
+	}
+	resolvedOrigin.Platform.OS = base.OS
+	resolvedOrigin.Platform.Channel = base.Channel.String()
+
+	return resultURL, resolvedOrigin, nil
+}
+
+// getCharm returns the charm being deployed. Either it already has been
+// used once and we get the data from state. Or we get the essential metadata.
+func (v *deployFromRepositoryValidator) getCharm(charmURL *charm.URL, resolvedOrigin corecharm.Origin) (corecharm.Origin, charm.Charm, error) {
+	repo, err := v.getCharmRepository(corecharm.CharmHub)
+	if err != nil {
+		return resolvedOrigin, nil, err
+	}
+
+	// Check if a charm doc already exists for this charm URL. If so, the
+	// charm has already been queued for download so this is a no-op. We
+	// still need to resolve and return back a suitable origin as charmhub
+	// may refer to the same blob using the same revision in different
+	// channels.
+	//
+	// We need to use GetDownloadURL instead of ResolveWithPreferredChannel
+	// to ensure that the resolved origin has the ID/Hash fields correctly
+	// populated.
+	// TODO: Handle already deployed charm.
```

Comment:
This needs to be figured out ☝️ Not sure what's the best way to see if a charm url is already deployed.

Reply:
The best way is the commented out code directly beyond this todo, check to see if the charm as been downloaded, thus existing in state for the current model. That's done by the charmdownloader today. Have to find a resolved charm origin as well, even if the charm has already been downloaded.

There are details of getting the charm object into the correct format etc.  So I stopped with the TODO. It'd be easy to keep going on all the parts, and keep making the PR bigger. :-)